src/test/java/guide/UsingHttpsTest.java

```diff
@@ -56,29 +68,133 @@ public void shouldPassWhenExpectedStubDidHappen() throws GeneralSecurityExceptio
     }
 
+    @Test
+    public void shouldFailSendingNonAllowedCertificate() throws IOException {
+        // client authentication
+        server.clientAuth(Files.readAllBytes(Path.of("build/resources/test/keystore_server_test_cert")), "changeit").run();
+        whenHttp(server).match(get("/asd")).then(ok()).mustHappen(0);
+
+        // sending a client certificate that is not allowed
+        InputStream clientKeyStore = StubServer.class.getResourceAsStream("keystore_server");
+        String clientKeyStorePass = "secret";
+
+        try {
+            sslReadyHttpAuthClient(clientKeyStore, clientKeyStorePass).execute(new HttpGet("https://localhost:" + server.getPort() + "/asd"));
+            fail("Request should have failed as server does not accept this client certificate");
+        } catch (Exception e) {
+            assertThat(e, isA(IOException.class));
+        }
+
+        ensureHttp(server).gotStubsCommitmentsDone();
+    }
+
     /**
      * Helper which returns HTTP client configured for https session
      */
     private HttpClient sslReadyHttpClient() throws GeneralSecurityException {
-        final SSLContext context = new SSLContextBuilder().loadTrustMaterial(null, new TrustSelfSignedStrategy()).build();
+        return sslReadyHttpClient(StubServer.getTrustStore(), null, null);
+    }
+
+    /**
+     * Helper which returns HTTP client configured for https session with custom trust store
+     */
+    private HttpClient sslReadyHttpClient(KeyStore trustStore) throws GeneralSecurityException {
+        return sslReadyHttpClient(trustStore, null, null);
+    }
+
+    /**
+     * Helper which returns HTTP client configured for https session with client authentication
+     */
+    private HttpClient sslReadyHttpAuthClient(InputStream keyStore, String keyStorePass) throws GeneralSecurityException {
+        return sslReadyHttpClient(StubServer.getTrustStore(), keyStore, keyStorePass);
+    }
+
+    /**
+     * Helper which returns HTTP client configured for https session
+     *
+     * @param trustStore trust store containing server's certificate
+     * @param keyStore key store for client authentication
+     * @param keyStorePass key store password
+     */
+    private HttpClient sslReadyHttpClient(KeyStore trustStore,
+                                          InputStream keyStore, String keyStorePass) throws GeneralSecurityException {
+        SSLContextBuilder sslContextBuilder = SSLContexts.custom();
+
+        // verifying server's identity
+        if (trustStore != null) {
+            sslContextBuilder.loadTrustMaterial(trustStore, null);
+        }
+
+        // client authentication
+        if (keyStore != null) {
+            try {
+                KeyStore store = KeyStore.getInstance(KeyStore.getDefaultType());
+                store.load(keyStore, keyStorePass.toCharArray());
+                sslContextBuilder.loadKeyMaterial(store, keyStorePass.toCharArray());
```

Comment:
I tried to simplify this a bit and reuse existing code for loading KS

Reply:
Looks good, thanks.